### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by narayanprusty**

### CODE REVIEW

File: docs/devnet.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+## Deploying a Devnet
+
+This document explains the flow of deploying a devnet and web services.
+
+### Authentication
+
+First you need to set up authentication to GCP. Download spacecraft-id-f8aeb7a6d337.json from 1password and place it in ~ and then set the following ENVs:
```
Comment (moshababo) [inline docs/devnet.md:7]:
```suggestion
First you need to set up authentication to GCP. 

```
We can keep the private info out.

File: docs/devnet.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+### Set Cloudflare API Token
+
+Set cloudflare API token in ENV so that all the services can be exposed publicly: 
+
+```
+export SPACECRAFT_CLOUDFLARE_API_TOKEN=
+```
+
+You can find the token in 1password
```
Comment (moshababo) [inline docs/devnet.md:30]: We can keep the private info out.

File: docs/devnet.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+### Slack Alerts
+
+To enable slack alerts you need to set the following environment variables:
+
+```
+export SPACECRAFT_SLACK_CHANNEL_ID=
+export SPACECRAFT_SLACK_TOKEN=
+```
+
+You can find the slack token in 1password and channel ID is the channel where you want to publish the alerts to.
```
Comment (moshababo) [inline docs/devnet.md:82]:
```suggestion
Channel ID is the channel where you want to publish the alerts to.

```
We can keep the private info out.

File: docs/devnet.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+### Deploy Web Services
+
+To deploy web services first you need to download tls.key and tls.crt files from 1password and place it in ./artifacts/ws directory.
```
Comment (moshababo) [inline docs/devnet.md:86]:
```suggestion
To deploy web services first you first need to place the secret `tls.key` and `tls.crt` files in ./artifacts/ws directory.

```
Also, plz give some more info about these files. How were they created, why are they needed, etc.

File: docs/devnet.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+Username: admin
+Password: prom-operator
+
+### Smapp Config
+
+When you deploy a network using go-spacecraft it creates and uploads a config file to GCP storage. You can access it using https://console.cloud.google.com/storage/browser/spacecraft-data/devnet-archive
+
+### Slack Alerts
+
+To enable slack alerts you need to set the following environment variables:
```
Comment (moshababo) [inline docs/devnet.md:75]: Plz link this somehow to the "Monitoring and Alerts" clause in `README.md`.

File: README.md [hunk 6]
```diff
@@ -1,50 +1,129 @@
...
-./go-spacecraft createNetwork --config=./artifacts/devnet/config.json
+
+Now you can explore the CLI flags and sub-commands using:
+
+```jsx
+./go-spacecraft --help
 ```
 
-This command will take around ~10min to finish and at the end you should see the kibana URL. The generated config file is uploaded to GCP storage. 
+Now you can refer to [devnet deployment guide](docs/devnet.md) to understand how to use the CLI to deploy a new network and web services.
```
Comment (moshababo) [inline README.md:68]: Plz give some info about what a devnet is, how is it differs from a testnet (if at all), etc. Obviously only in the context of Spacecraft.

File: README.md [hunk 7]
```diff
@@ -1,50 +1,129 @@
...
+
+## Network Deployment
+
+For creating a new network boostrap deploys one bootstrap node, multiple bootnodes and then multiple miners. The bootnodes connect to bootstrap for p2p discovery and miners connect to bootnodes for p2p discovery.
```
Comment (moshababo) [inline README.md:72]: Try to give more details here, to make it clearer.

File: README.md [hunk 8]
```diff
@@ -1,50 +1,129 @@
...
+Now you can refer to [devnet deployment guide](docs/devnet.md) to understand how to use the CLI to deploy a new network and web services.
+
+## Network Deployment
+
+For creating a new network boostrap deploys one bootstrap node, multiple bootnodes and then multiple miners. The bootnodes connect to bootstrap for p2p discovery and miners connect to bootnodes for p2p discovery. 
+
+Spacecraft also allows you to create a network that actually extends another network i.e., deploys a network without bootstrap. In this case the bootnodes connect to the bootnodes of the other network and miners of this new network connect to the bootnodes of this new network. This can be useful when you want to delete all the managed miners and other deployments and create a fresh network that connects to an existing network. You can achieve this using the `--bootstrap=false` flag.
+
+Whenever a network is created it also deploys a N number of PoETs. The PoET are assigned to the nodes in round robin fashion. The first node is then added to the config file so home smeshers use the first PoET always.
```
Comment (moshababo) [inline README.md:76]: How is this defined? What's the default N?

Comment (moshababo) [inline README.md:76]: How are nodes being assigned to the PoETs, as their gateway nodes?

File: README.md [hunk 9]
```diff
@@ -1,50 +1,129 @@
...
+
+The bootnodes are always tied to a particular k8s worker node in round robin fashion. There are mainly two reasons for doing this: 
+
+- Firstly if we don't do this then its possible k8s will host multiple bootnodes in one k8s worker node and if that VM crashes then all bootnodes are down and new smeshers cannot join the network.
+- Secondly the TCP port of the bootnodes needs to be bind'ed to the host node otherwise p2p discovery fails. p2p discover doesn't work over k8s NodePort service. Therefore if we don't bind bootnodes to host node then the k8s node in which the bootnodes are hosted can be rescheduled and therefore the IP to reach the bootnodes will change and smeshers will not be able to connect to the all or some bootnodes.
+
+Spacecraft calculates the total number of k8s nodes need to be created dynamically based on total pods and their resource size. 
+
+The GRPC and JSON API ports of all the managed miners are exposed as NodePort. You can get list of all the managed miners GRPC URLs using the `hosts` sub-command.
```
Comment (moshababo) [inline README.md:85]: If these are 2 different ports, how are they exposed as a single `NodePort`?

File: README.md [hunk 10]
```diff
@@ -1,50 +1,129 @@
...
+
+## Logs
+
+Spacecraft deploys ELK stack for aggregation of logs. It uses filebeat to collect logs and directly stores them in Elasticsearch. All the fields in the logs are converted to string and non-JSON logs are stored raw. There is no logstash deployed because logstash is slow at processing logs instead it uses filebeat logs processing scripts which can process logs  in parallel and very fast.
+
+To deploy filebeat, elasticsearh and kibana it uses helm charts by elastic ([https://github.com/elastic/helm-charts](https://github.com/elastic/helm-charts)).
+
+## Monitor and Alerts
```
Comment (moshababo) [inline README.md:93]: What about giving some info about Pyroscope?

File: README.md [hunk 11]
```diff
@@ -1,50 +1,129 @@
...
+
+Cloudflare is used as SSL proxy and for domain records for spacemesh public JSON API, explorer backend API and dashboard backend API. Whereas for spacemesh public GRPC API, k8s Ingress  is used as SSL proxy and cloudflare is used only for DNS record. The reason we do this for GRPC API is because cloudflare is capable to proxy GRPC connections currently. Due to this we store SSL cert and key in secrets.
+
+## Metrics
+
+If the `--metrics` flag is set during network deployment then prometheus and grafana is deployed to collect metrics. Prometheus and grafana is deployed using helm charts. These are deployed in `metrics` namespace. 
+
+Here is the architecture of the deployment:
+
+![metrics.png](docs/metrics.png)
```
Comment (moshababo) [inline README.md:127]: These components are not described in `deployment_arch.png`. It should be added if possible. 
Also, are you sure that all of them are implemented?

**Review by moshababo — COMMENTED** [review]
Looks good!

File: README.md [hunk 12]
```diff
@@ -1,50 +1,129 @@
...
+
+Here is high level architecture diagram of the dashboard backend components and their interactions:
+
+![dash_arch_chart.png](docs/dash_arch_chart.png)
+
+Here is high level architecture diagram of the explorer backend components and their interactions:
+
+![explorer_arch_chart.png](docs/explorer_arch_chart.png)
+
+Cloudflare is used as SSL proxy and for domain records for spacemesh public JSON API, explorer backend API and dashboard backend API. Whereas for spacemesh public GRPC API, k8s Ingress  is used as SSL proxy and cloudflare is used only for DNS record. The reason we do this for GRPC API is because cloudflare is capable to proxy GRPC connections currently. Due to this we store SSL cert and key in secrets.
```
Comment (avive) [inline README.md:119]: I think you mean `currently not capable` if I understand this issue correctly.

**Review by avive — CHANGES_REQUESTED** [review]
This reads great. Had one minor comment.

File: README.md [hunk 13]
```diff
@@ -1,50 +1,129 @@
...
+Spacecraft is a CLI tool which lets us deploy Spacemesh networks in GCP. It uses kubernetes REST APIs and Helm charts for deployment of different services. 
 
-The default configuration of the tool lets you launch a mininet i.e., a network consisting of 10 miners and 1 poet. Mininet is helpful for development and testing of spacecraft tool itself but to test go-spacemesh or poet builds you need to launch a devnet as it has configured to enable to network run for long time.
+It allows you to create a new network, add miner to existing network, delete a miner from an network, upgrade config or spacemesh release of a network, create a network network without bootstrap, calculate rewards of a network, deploy web services for a network.
 
-## Launching a Devnet
+Here is an high level architecture of a complete network deployed on GCP using spacecraft:
 
-To launch a devnet follow the below steps:
+![deployment_arch.png](docs/deployment_arch.png)
```
[the remark this answers is not in the document]
Comment (narayanprusty) [reply README.md:9]: It's in the docs/spacecraft.gliffy

File: README.md (hunk 11 above)
Comment (narayanprusty) [reply README.md:127]: Yes they all are implemented. It's deployed in a separate namespace therefore I think it's fine to show it as a seperate architecture diagram.

File: README.md (hunk 12 above)
Comment (narayanprusty) [reply README.md:119]: Yes "not capable". Thanks for catching it.

File: README.md (hunk 8 above)
Comment (moshababo) [reply README.md]: Was this resolved? Still can't find info about it.

File: README.md (hunk 9 above)
Comment (moshababo) [reply README.md]: Was this resolved? Still can't find info about it.

File: README.md [hunk 14]
```diff
@@ -1,50 +1,141 @@
...
+
+To deploy filebeat, elasticsearh and kibana it uses helm charts by elastic ([https://github.com/elastic/helm-charts](https://github.com/elastic/helm-charts)).
+
+## Monitor and Alerts
+
+Spacecraft deploys spacemesh-watch ([https://github.com/spacemeshos/spacemesh-watch](https://github.com/spacemeshos/spacemesh-watch)) service which uses GRPC API of the managed miners to monitor them. And when some issue is detected it send alerts in slack. 
+
+## Pyroscope
+
+Pyroscope is an continuous profiling platform. Spacecraft also deploys pyroscope for debugging performance related issues of miners. Pyroscope doesn't support horizontal scaling therefore we cannot collect data of all the miners so we only collect data of miner-10 and miner-20.
```
Comment (moshababo) [inline README.md:107]: Is the `miner-10`, `miner-20` selection static/hardcoded?

File: README.md [hunk 15]
```diff
@@ -1,50 +1,129 @@
...
+![deployment_arch.png](docs/deployment_arch.png)
 
-### Install Go
+## Service Integrations
 
-This tool requires go 1.15 or above installed. If you haven't installed golang then follow the instructions here https://golang.org/doc/install
+Spacecraft integrates with mainly 3 different services: GCP, Cloudflare and Slack. In GCP it uses, kubernetes engine, cloud storage and load balancer. 
 
-### Set secrets in ENV
+It creates a new kubernetes cluster for every new network and stores the network go-sm config file in cloud storage.
```
[the remark this answers is not in the document]
Comment (moshababo) [reply README.md:23]: This wasn't applied.

File: README.md [hunk 16]
```diff
@@ -1,50 +1,129 @@
...
-Set the following environment variables:
+The discovery service ([https://discover.spacemesh.io/networks.json](https://discover.spacemesh.io/networks.json)) is already hosted on cloud storage and served as static site through load balancer. For every new network deployment, spacecraft only updates the discovery service networks list file stored in cloud storage.
 
-```
-SPACECRAFT_GCP_PROJECT=...
-SPACECRAFT_GCP_LOCATION=...
-SPACECRAFT_GCP_ZONE=...
-GOOGLE_APPLICATION_CREDENTIALS=...
-```
+Cloudflare is used for creating domain name records for every new network. the records point to various services deployed in kubernetes engine. Our cloudflare account is already configured with [spacemesh.io](http://spacemesh.io) nameservers therefore spacecraft can create the sub-domain records in it. Spacecraft currently doesn't allow us to provide root domain name as CLI option. It has spacemesh.io root domain hardcoded.
```
[the remark this answers is not in the document]
Comment (moshababo) [reply README.md:27]: What about this?

[the remark this answers is not in the document]
Comment (moshababo) [reply README.md:27]: ☝️

File: README.md (hunk 8 above)
Comment (narayanprusty) [reply README.md]: Yes its resolved. I had added this line "The PoET are assigned to the nodes in round robin fashion using `poet-server` config"

File: README.md (hunk 9 above)
Comment (narayanprusty) [reply README.md]: Yes its resolved. I had added this line "The GRPC and JSON API ports of all the managed miners are exposed as separate NodePort. You can get list of all the managed miners GRPC URLs using the `hosts` sub-command."

File: README.md (hunk 14 above)
Comment (narayanprusty) [reply README.md]: Yes hardcoded. I have updated the readme.

File: docs/devnet.md (hunk 4 above)
Comment (narayanprusty) [reply docs/devnet.md]: Added more information about it

File: README.md (hunk 7 above)
Comment (narayanprusty) [reply README.md:95]: I am not sure what else to be added here.

**Review by moshababo — APPROVED** [review]

**Pull request merged**